src/main/java/dk/kb/util/yaml/YAML.java

```diff
@@ -668,52 +684,293 @@ public boolean containsKey(Object path) throws NullPointerException {
+                } else {
+                    visitor.visit(entry.getValue());
+                }
+            }
+        } else if (conditionalMatch.matches()) {
+            conditionalTraverse(yPath, visitor, conditionalMatch, map, shortenedPath);
+        } else {
+            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                String key = entry.getKey().toString();
+                Object value = entry.getValue();
+
+                if (yPath.size() == 1){
+                    if (key.equals(yPath.get(0))){
+                        visitor.extractedValues.add(value);
+                    }
+                } else {
+                    if (key.equals(yPath.get(0))){
+                        traverse(shortenedPath, value, visitor);
+                    }
+                }
             }
-            Object sub;
-            if (pathKey.isEmpty()) {
-                sub = current;
-            } else {
-                sub = current.getSuper(pathKey);
+        }
+    }
+
+    private void conditionalTraverse(List<String> yPath, MultipleValuesVisitor visitor, Matcher conditionalMatch, Map<?, ?> map, List<String> shortenedPath) {
+        String key = conditionalMatch.group(1);
+        boolean mustMatch = conditionalMatch.group(2).equals("="); // The Pattern ensures only "!=" or "=" is in the group
+        String value = conditionalMatch.group(3);
+
+        List<Object> matchingObjects = new ArrayList<>();
+        for (Map.Entry<?,?> entry : map.entrySet()) {
+            matchingObjects.add(conditionalGet(entry.getValue(), key, value, mustMatch));
+        }
+
+        matchingObjects = matchingObjects.stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        if (yPath.size() <= 1){
+            for (Object matchingEntry: matchingObjects) {
+                visitor.visit(matchingEntry);
             }
-            if (sub == null) {
-                throw new NotFoundException(
-                        "Unable to request " + i + "'th sub-element: '" + pathKey + "'", path);
+        } else {
+            for (Object matchingEntry: matchingObjects) {
+                traverse(shortenedPath, matchingEntry, visitor);
             }
+        }
+    }
+
+
+    /**
+     * If the first element in a list starts with '[' and ends with ']' remove the brackets from the string. The brackets
+     * are used to describe a conditional lookup in a YAML map and needs to be removed before the traversal can be
+     * continued.
+     * @param yPath list of path elements, where the first element is the current element of the YAML traversal, which
+     *              is checked for '[' and ']'.
+     */
+    private static void removeBracketsFromPathElement(List<String> yPath) {
+        if (!yPath.isEmpty() && yPath.get(0).startsWith("[") && yPath.get(0).endsWith("]")){
+            String cleaned = yPath.get(0).substring(1, yPath.get(0).length()-1);
+            yPath.set(0, cleaned);
```

Comment:
The Pattern `ARRAY_ELEMENT` can be used for this and makes for a cleaner implementation.

Reply:
Nearly there :-)

```Java
    String cleaned = yPath.get(0).substring(1, yPath.get(0).length()-1);
```
can be replaced with
```Java
    String cleaned = arrayMatcher.group(2);
```